src/Symfony/Component/Mailer/Transport/Dsn.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Mailer\Transport;
+
+use Symfony\Component\Mailer\Exception\InvalidArgumentException;
+
+/**
+ * @author Konstantin Myakshin <molodchick@gmail.com>
+ */
+final class Dsn
+{
+    private $scheme;
+    private $host;
+    private $user;
+    private $password;
+    private $port;
+    private $options;
+
+    public function __construct(string $scheme, string $host, ?string $user = null, ?string $password = null, ?int $port = null, array $options = [])
+    {
+        $this->scheme = $scheme;
+        $this->host = $host;
+        $this->user = $user;
+        $this->password = $password;
+        $this->port = $port;
+        $this->options = $options;
+    }
+
+    public static function fromString(string $dsn): self
+    {
+        if (false === $parsedDsn = parse_url($dsn)) {
+            throw new InvalidArgumentException(sprintf('The "%s" mailer DSN is invalid.', $dsn));
+        }
+
+        if (!isset($parsedDsn['scheme'])) {
+            throw new InvalidArgumentException(sprintf('The "%s" mailer DSN must contain a transport scheme.', $dsn));
+        }
+
+        if (!isset($parsedDsn['host'])) {
+            throw new InvalidArgumentException(sprintf('The "%s" mailer DSN must contain a mailer name.', $dsn));
+        }
+
+        $user = urldecode($parsedDsn['user'] ?? null);
+        $password = urldecode($parsedDsn['pass'] ?? null);
```

Comment:
this is broken. It is passing `null` to `urldecode`, and it won't get `null` as `$user` then.

Reply:
see 288772478cb4d96d340bc946745a859a0b01334a

Reply:
But it already covered with test https://github.com/symfony/symfony/pull/31946/files#diff-e3a1e78245b60fed36ed6ab4e56023ceR48 . Strange. Maybe because `assertEquals` compares `''` equal to `null`

Reply:
@Koc `assertEquals()` does not make a strict comparison unlike `assertSame()`. It is therefore recommended to use `assertSame()` (or `assertNull()` to check the null value).